libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp

```diff
@@ -114,15 +130,50 @@ void AP_BattMonitor_UAVCAN::handle_battery_info(const BattInfoCb &cb)
     _interim_state.healthy = true;
 }
 
+void AP_BattMonitor_UAVCAN::handle_cbat(const CBattCb &cb)
+{
+    WITH_SEMAPHORE(_sem_battmon);
+    _interim_state.voltage = cb.msg->voltage;
+    _interim_state.current_amps = cb.msg->current;
+    _interim_state.time_remaining = cb.msg->average_time_to_empty * 60;
+    _interim_state.consumed_mah = cb.msg->passed_charge * 1000;
+    _soc = cb.msg->state_of_charge;
+    _cycles = cb.msg->cycle_count;
+
+    if (!isnanf(cb.msg->temperature) && cb.msg->temperature > 0) {
+        // Temperature reported from battery in kelvin and stored internally in Celsius.
+        _interim_state.temperature = cb.msg->temperature - C_TO_KELVIN;
+        _interim_state.temperature_time = AP_HAL::millis();
+    }
+
+    for (uint8_t i = 0; i < cb.msg->cell_count; i++) {
+        _interim_state.cell_voltages.cells[i] = cb.msg->voltage_cell[i] * 1000;
+    }
```

Comment:
This is a memory over-write waiting to happen; you need to constrain your loop to `ARRAY_SIZE(_interim_state.cell_voltages.cells[i])` as well as `cb.msg->cell_count`